### PR TITLE
Refactor `authUser` to avoid nested `runPersistent` calls

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR refactors the `authUser` function to avoid nested calls which would cause the program to hang.

### CODE REVIEW

File: users-persistent/src/Web/Users/Persistent.hs [hunk 1]
```diff
@@ -188,16 +189,12 @@ instance UserStorageBackend Persistent where
...
-        runPersistent conn $
-        do mFst <- selectFirst ([LoginUsername ==. userOrEmail] ||. [LoginEmail ==. userOrEmail]) []
-           case mFst of
-             Nothing -> return Nothing
-             Just login ->
-                 do user <- unpackLogin' (entityVal login)
-                    if authFn user
-                    then liftIO $ Just <$> action (entityKey login)
-                    else return Nothing
+    withAuthUser conn userOrEmail authFn action = runMaybeT $ do
```
Comment (agrafix) [inline users-persistent/src/Web/Users/Persistent.hs]: i'd prefer something like

``` haskell
withAuthUser conn userOrEmail authFn action =
    runMaybeT $
    do login <- 
             MaybeT $ liftIO $ runPersistent conn $ selectFirst ...
       user <- unpackLogin' $ entityVal login
       guard $ authFn user
       liftIO . action . entityKey $ login
```

**Comment by agrafix** [comment]
Thank's for the patch! Looks good. Three suggestions:
- I've made a suggestion about formatting for the function to keep it in sync with the rest of the project. What do you think?
- could you remove the whitespace changes?
- do you think it would be easy to write a unit test to make sure this does not happen again?

Sadly travis fails due to another reason :-( I'll look into that!

**Comment by parsonsmatt** [comment]
Done!

I'm not sure about how I'd write a unit test for the behavior -- the library passes the current test suite, which should have caused the problem with the calls to `authUser`. It may be a Postgres/Persistent specific thing.

**Comment by agrafix** [comment]
Great thank you! I'll merge this PR for now and have a look at it when I fix the travis build. Do you need ghc7.6 support?

**Comment by parsonsmatt** [comment]
Thanks! :smile: 

I'm lucky enough to be on the latest GHC. Is there something I can do to help support the older version?

**Comment by agrafix** [comment]
No I think we can drop GHC7.6 support then. My projects are on GHC7.10 too. 

Would you like write access to the repo to help out maintaining?

**Comment by parsonsmatt** [comment]
Unfortunately, I don't think I will have much time in the next few months to help out that actively. I might reconsider in May when I'm graduated

**Pull request merged**